### DISCUSSION TITLE
Manual "new"

### DISCUSSION THREAD

**Discussion by spoerri** [jj-vcs/jj#4047] [Ideas]
On the back of #4012, it'd be nice to reduce the dependency on empty commits. Here's a different way to achieve that:

1. introduce a command `jj view` that loads the specified revision into the working copy, and makes all its files and directories readonly.
2. when abandoing a commit, delete that commit. if the commit is currently associated with the working copy, then use the same approach as `jj view`, i.e. make the working copy readonly.

**Comment by emilazy** [comment]
How can you make the working copy read‐only? Jujutsu can’t stop files from changing underneath it (unless you think we should require use of a VFS). Would it just throw an error and refuse to run if any of them differ?

**Reply by spoerri** [reply]
I guess I was assuming unix-likes. In windows, we would only be able to make the existing files readonly. And, yes could fall back to complaining in that case and if the user changes permissions back to readwrite.

**Reply by emilazy** [reply]
I don’t understand how you could accomplish this on Unix either.

**Reply by emilazy** [reply]
Ah, I guess you just mean removing the write permissions from files? It’d also have to touch the directory containing the repository, which is kind of awkward. What would you do to restore the pristine state if a tool changes the permissions and modifies a file?

**Reply by spoerri** [reply]
That's right, remove write permissions from files and directories, including the top one, which should be fine, because the .jj, .git, etc directories already exist and wouldn't be involved.

> What would you do to restore the pristine state if a tool changes the permissions and modifies a file?

Well, if a person doesn't know what's going on, and does wrongfully `chmod u+w someFile; echo foo>>someFile; jj st` or really any jj command, then it should print something like:
> Error! There are writable files in what is supposed to be a readonly jj working copy. If you really would like to modify a jj change, please run jj edit or jj new. If you'd prefer just view the change, run jj view, and jj will change the files to not be writable again. Note that running jj view will delete any changes you've made in the working copy.

...and not do anything at all, including not modifying the current change.

Did I understand the question? I'm having trouble thinking of tools that would go and make files writable without permission, which seems unfriendly. Perhaps vim toggles the write bit of its swap files under some circumstances?

Ah maybe you mean files that are "owned" by a program, such that the program makes them readonly to discourage users from modifying them without using the program? Well, IIUC git doesn't support storing those permissions anyway, so if those files are being stored in the repo, people are dealing with this already. Perhaps the error message should mention something about `jj untrack` and list some of the writable files to be helpful.

**Reply by joyously** [reply]
Wouldn't it be much easier to leave the working copy alone and output the historical version somewhere else, like a folder named for change ID?  If you aren't "working" on it (just viewing), why affect the working copy?

**Reply by spoerri** [reply]
Do you mean easier for the user in case of this error, jj in the case of this error, or in general?

If you mean in general, that'd be a very big change to jj design and UI. One could suggest that, but I'm not 😄

**Reply by joyously** [reply]
I meant easier for the user to utilize the historical version separately from the working copy, easier for `jj` to ensure that the working copy is unaffected without having to make it read-only, easier for the user to not have to fight with the tool about read-only files, easier to not create a situation that is an error when it really isn't an error, and easier in general for users to keep historical versions separate from the working copy when just wanting to view them (especially to be able to see both at once).
This proposal is for a new command `view`, so it wouldn't be changing `jj` design since it doesn't exist.

**Reply by spoerri** [reply]
Wouldn't help with point 2 of the top idea

**Reply by arxanas** [reply]
You could make a working copy read-only *from the perspective of jj* by simply passing the equivalent of `--ignore-working-copy` to every command, when some configuration is in effect. Then it'll look like the working copy never changes, as far as jj knows. Combined with ideas like "lazy-new", then it means that there's a mode where jj wouldn't automatically create new working-copy commits. But I don't know if that's what's meant by the discussion or if that would help the intended workflow.

**Reply by spoerri** [reply]
> --ignore-working-copy

that's cool!

indeed it does not serve my fell purpose of killing the concept of a "finished" commit and aggressive empty commits, coming up again in [the advance-branches discussion](https://github.com/martinvonz/jj/discussions/3549#discussioncomment-9910893) 

the fundamental challenge is that the user always wants a working copy for which we will "do the right thing". mostly that means no editing finished commits by default. either lazy-new or manual-new (leveraging a readonly working copy) would address that challenge. both lazy-new and manual-new allow only having a new commit when you actually want one, and not for the sake of finishing some other commit.

instead disconnecting the working copy from any change would be worse for understandability.

**Reply by spoerri** [reply]
you know, maybe this is all overkill, and it'd be practically sufficient to just remove/deprecate `jj commit`

that is, remove the expectation that you have to do anything when you're done working on a change. that would mean you always use \@ for your work, and never \@-. so `jj describe -m foo; jj git push` is best. even though `jj commit -m foo; jj git push -c @-` would work in the short term, it doesn't have to be recommended.

**Reply by arxanas** [reply]
@spoerri does jj actually have an expectation of using `jj commit`? I didn't see it suggested in the docs. It's one of those "comfort" commands for people from other VCSes, mainly, so I didn't think we explicitly prescribed it for workflows.

**Reply by spoerri** [reply]
You could argue it’s just for GitHub interoperability and maybe for advance-branches.

**Reply by arxanas** [reply]
I don't think `jj commit` particularly helps with GitHub interoperability, at least not more than `jj describe` does? The restriction AIUI is just that you can't push commits without descriptions.

I don't see a reason to remove/deprecate `jj commit`; whom does that help? (Also, if it gets special advance-branches treatment, then it probably improves the case for _not_ removing it.)

**Reply by spoerri** [reply]
Sorry, was abbreviating re: needing to care about @-, needing to have empty commits in order for things to work, and using `jj commit` to finish work: I have seen a sentiment that these are only a problem for using git forges.

For example, will certainly work with github if you do `jj new; echo bar > baz; jj commit -m foo; jj git push -c @-` and wishing that it would work with even `jj new -m foo; echo bar > baz; jj push` (notice the the actual change happens after the "describe").

It's unfortunate to need users to run both "new" before their changes and "commit" after their changes. IMHO eventually `jj git push` should just take care of it: update the branch from the working copy and push it.

**Reply by arxanas** [reply]
> wishing that it would work with even jj new -m foo; echo bar > baz; jj push

I don't understand why this doesn't work now? It seems to work for me if I run `jj new -m foo` and then `jj git push`, because the current commit has a non-empty message.

**Reply by spoerri** [reply]
i guess it's really just the branches. would like them to automatically advance on `jj git push`
